node/node_v2.go

```diff
@@ -0,0 +1,126 @@
+	quorum         core.QuorumID
+}
+type relayRequest struct {
+	chunkRequests []*clients.ChunkRequestByRange
+	metadata      []*requestMetadata
+}
+type response struct {
+	metadata []*requestMetadata
+	bundles  [][]byte
+}
+
+type RawBundles struct {
+	BlobCertificate *corev2.BlobCertificate
+	Bundles         map[core.QuorumID][]byte
+}
+
+func (n *Node) DownloadBatch(ctx context.Context, batch *corev2.Batch) ([]*corev2.BlobShard, []*RawBundles, error) {
+	if n.RelayClient == nil {
+		return nil, nil, fmt.Errorf("relay client is not set")
+	}
+
+	operatorState, err := n.ChainState.GetOperatorStateByOperator(ctx, uint(batch.BatchHeader.ReferenceBlockNumber), n.Config.ID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	blobShards := make([]*corev2.BlobShard, len(batch.BlobCertificates))
+	rawBundles := make([]*RawBundles, len(batch.BlobCertificates))
+	requests := make(map[corev2.RelayKey]*relayRequest)
+	for i, cert := range batch.BlobCertificates {
+		blobKey, err := cert.BlobHeader.BlobKey()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get blob key: %v", err)
+		}
+
+		if len(cert.RelayKeys) == 0 {
+			return nil, nil, fmt.Errorf("no relay keys in the certificate")
+		}
+		blobShards[i] = &corev2.BlobShard{
+			BlobCertificate: cert,
+			Bundles:         make(map[core.QuorumID]core.Bundle),
+		}
+		rawBundles[i] = &RawBundles{
+			BlobCertificate: cert,
+			Bundles:         make(map[core.QuorumID][]byte),
+		}
+		relayIndex := rand.Intn(len(cert.RelayKeys))
+		relayKey := cert.RelayKeys[relayIndex]
+		for _, quorum := range cert.BlobHeader.QuorumNumbers {
+			assgn, err := corev2.GetAssignment(operatorState, batch.BlobCertificates[0].BlobHeader.BlobVersion, quorum, n.Config.ID)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to get assignments: %v", err)
+			}
+
+			req, ok := requests[relayKey]
+			if !ok {
+				req = &relayRequest{
+					chunkRequests: make([]*clients.ChunkRequestByRange, 0),
+					metadata:      make([]*requestMetadata, 0),
+				}
+				requests[relayKey] = req
+			}
+			// Chunks from one blob are requested to the same relay
+			req.chunkRequests = append(req.chunkRequests, &clients.ChunkRequestByRange{
+				BlobKey: blobKey,
+				Start:   assgn.StartIndex,
+				End:     assgn.StartIndex + assgn.NumChunks,
+			})
+			req.metadata = append(req.metadata, &requestMetadata{
+				blobShardIndex: i,
+				quorum:         quorum,
+			})
+		}
+	}
+
+	pool := workerpool.New(len(requests))
+	bundleChan := make(chan response, len(requests))
+	for relayKey := range requests {
+		relayKey := relayKey
+		req := requests[relayKey]
+		pool.Submit(func() {
+			bundles, err := n.RelayClient.GetChunksByRange(ctx, relayKey, req.chunkRequests)
+			if err != nil {
+				n.Logger.Errorf("failed to get chunks from relays: %v", err)
+				return
+			}
+			bundleChan <- response{
+				metadata: req.metadata,
+				bundles:  bundles,
+			}
+		})
+	}
+	pool.StopWait()
+
+	for i := 0; i < len(requests); i++ {
```

Comment:
The pool may send less than `len(requests)` responses to the channel as it exits on error (L103), so this may hang/stuck the server.

Reply:
Thanks for catching this. Fixed!